{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.8.3\n"
     ]
    }
   ],
   "source": [
    "from platform import python_version\n",
    "\n",
    "print(python_version())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_sequence</th>\n",
       "      <th>protein_length</th>\n",
       "      <th>cdna</th>\n",
       "      <th>5utr</th>\n",
       "      <th>cds</th>\n",
       "      <th>3utr</th>\n",
       "      <th>cdna_length</th>\n",
       "      <th>5utr_length</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>3utr_length</th>\n",
       "      <th>protein_abundance</th>\n",
       "      <th>median_TPM</th>\n",
       "      <th>protein_per_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000341562</td>\n",
       "      <td>ENST00000341156</td>\n",
       "      <td>ENSG00000135454</td>\n",
       "      <td>MWLGRRALCALVLLLACASLGLLYASTRDAPGLRLPLAPWAPPQSP...</td>\n",
       "      <td>533</td>\n",
       "      <td>GCATTCCCCGCGCGGAGCCGAAGCAGCCGCAACGAGCCGGGAGCTG...</td>\n",
       "      <td>GCATTCCCCGCGCGGAGCCGAAGCAGCCGCAACGAGCCGGGAGCTG...</td>\n",
       "      <td>ATGTGGCTGGGCCGCCGGGCCCTGTGCGCTCTGGTCCTTCTGCTCG...</td>\n",
       "      <td>TGGCCCGCTGGGGATTTCTGACTGTCAGGCTGGGCCTGCCTCCTTG...</td>\n",
       "      <td>5368</td>\n",
       "      <td>431</td>\n",
       "      <td>1602</td>\n",
       "      <td>3335</td>\n",
       "      <td>0.068</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.680000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000290765</td>\n",
       "      <td>ENST00000290765</td>\n",
       "      <td>ENSG00000133433</td>\n",
       "      <td>MGLELFLDLVSQPSRAVYIFAKKNGIPLELRTVDLVKGQHKSKEFL...</td>\n",
       "      <td>244</td>\n",
       "      <td>AGCTGCTGCCCACACCGCGCTCAGCGCCTTCACTGCCATCCCCGCT...</td>\n",
       "      <td>AGCTGCTGCCCACACCGCGCTCAGCGCCTTCACTGCCATCCCCGCT...</td>\n",
       "      <td>ATGGGCCTAGAGCTGTTTCTTGACCTGGTGTCCCAGCCCAGCCGCG...</td>\n",
       "      <td>AGGGTCTGGGATGGGGGCCAGGAGATTAGCAACAAGGATTCATTCT...</td>\n",
       "      <td>1108</td>\n",
       "      <td>64</td>\n",
       "      <td>735</td>\n",
       "      <td>309</td>\n",
       "      <td>57.000</td>\n",
       "      <td>11.385</td>\n",
       "      <td>5.006588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000256996</td>\n",
       "      <td>ENST00000256996</td>\n",
       "      <td>ENSG00000134574</td>\n",
       "      <td>MAPKKRPETQKTSEIVLRPRNKRSRSPLELEPEAKKLCAKGSGPSR...</td>\n",
       "      <td>427</td>\n",
       "      <td>GGTTTGAACAAGCCCTGGGCATGTTTGGCGGGAAGTTGGCTTAGCT...</td>\n",
       "      <td>GGTTTGAACAAGCCCTGGGCATGTTTGGCGGGAAGTTGGCTTAGCT...</td>\n",
       "      <td>ATGGCTCCCAAGAAACGCCCAGAAACCCAGAAGACCTCCGAGATTG...</td>\n",
       "      <td>GAGACACTAAAGAAGGTGTGGGCCAGACAAGGCCTTGGAGCCCACA...</td>\n",
       "      <td>1815</td>\n",
       "      <td>163</td>\n",
       "      <td>1284</td>\n",
       "      <td>368</td>\n",
       "      <td>2.510</td>\n",
       "      <td>3.740</td>\n",
       "      <td>0.671123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000262990</td>\n",
       "      <td>ENST00000262990</td>\n",
       "      <td>ENSG00000109445</td>\n",
       "      <td>MPKKKTGARKKAENRREREKQLRASRSTIDLAKHPCNASMECDKCQ...</td>\n",
       "      <td>320</td>\n",
       "      <td>GCTGTCGTAAAAGGACGTCCGGTCCGTCTCCTAGTGTCCGGAATCG...</td>\n",
       "      <td>GCTGTCGTAAAAGGACGTCCGGTCCGTCTCCTAGTGTCCGGAATCG...</td>\n",
       "      <td>ATGCCTAAAAAAAAGACTGGTGCGAGGAAGAAGGCTGAGAACCGCC...</td>\n",
       "      <td>GGGAGCTGCTCTGGTGGCCGTGTGTGAGAGGAGCAGGAGTGAGTGT...</td>\n",
       "      <td>1891</td>\n",
       "      <td>220</td>\n",
       "      <td>963</td>\n",
       "      <td>708</td>\n",
       "      <td>2.500</td>\n",
       "      <td>13.445</td>\n",
       "      <td>0.185943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000371811</td>\n",
       "      <td>ENST00000382374</td>\n",
       "      <td>ENSG00000165487</td>\n",
       "      <td>MAAAAGSCARVAAWGGKLRRGLAVSRQAVRSPGPLAAAVAGAALAG...</td>\n",
       "      <td>434</td>\n",
       "      <td>GCCTAGCTGCGCTTCCGCAAAGATGGCGGCGGCTGCGGGTAGCTGC...</td>\n",
       "      <td>GCCTAGCTGCGCTTCCGCAAAG</td>\n",
       "      <td>ATGGCGGCGGCTGCGGGTAGCTGCGCGCGGGTGGCGGCCTGGGGCG...</td>\n",
       "      <td>TAAAAGATATAATAGTATGGCAATTATATTGTTCCAAATGTCAAAA...</td>\n",
       "      <td>1885</td>\n",
       "      <td>22</td>\n",
       "      <td>1305</td>\n",
       "      <td>558</td>\n",
       "      <td>2.500</td>\n",
       "      <td>8.060</td>\n",
       "      <td>0.310174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7982</th>\n",
       "      <td>ENSP00000387111</td>\n",
       "      <td>ENST00000410042</td>\n",
       "      <td>ENSG00000273155</td>\n",
       "      <td>MAGILRLVVQWPPGRLQTVTKGVESLICTDWIRHKFTRSRIPEKVF...</td>\n",
       "      <td>131</td>\n",
       "      <td>GCACTTTTCCAGCTCGAGCCCTCACGAGGCCGTGGGTACGACCGGA...</td>\n",
       "      <td>GCACTTTTCCAGCTCGAGCCCTCACGAGGCCGTGGGTACGACCGGA...</td>\n",
       "      <td>ATGGCTGGGATTTTGCGCTTAGTAGTTCAATGGCCCCCAGGCAGAC...</td>\n",
       "      <td>TGTAATTCTAAATGCATTTTTCTTTTTTCGTTAATGTTCTGAAAAA...</td>\n",
       "      <td>669</td>\n",
       "      <td>207</td>\n",
       "      <td>396</td>\n",
       "      <td>66</td>\n",
       "      <td>1.010</td>\n",
       "      <td>0.155</td>\n",
       "      <td>6.516129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7983</th>\n",
       "      <td>ENSP00000477037</td>\n",
       "      <td>ENST00000553909</td>\n",
       "      <td>ENSG00000259171</td>\n",
       "      <td>MVMGLGVLLLVFVLGLGLTPPTLAQDNSRHL</td>\n",
       "      <td>31</td>\n",
       "      <td>ACACACTCACACAAGGACGCCAACCCCACCTAGATGCAAAGCAGGA...</td>\n",
       "      <td>ACACACTCACACAAGGACGCCAACCCCACCTAGATGCAAAGCAGGA...</td>\n",
       "      <td>ATGGTGATGGGCCTGGGCGTTTTGTTGTTGGTCTTCGTGCTGGGTC...</td>\n",
       "      <td>GATACTGATGGCTCTGCAGAGGACCCATTCATTGCTTCTGCTTTTG...</td>\n",
       "      <td>1498</td>\n",
       "      <td>190</td>\n",
       "      <td>96</td>\n",
       "      <td>1212</td>\n",
       "      <td>0.591</td>\n",
       "      <td>24.155</td>\n",
       "      <td>0.024467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7984</th>\n",
       "      <td>ENSP00000361601</td>\n",
       "      <td>ENST00000372523</td>\n",
       "      <td>ENSG00000168612</td>\n",
       "      <td>MLERLKAPWSAALQRKYFDLGIWTAPISPMALTMLNGLLIKDSSPP...</td>\n",
       "      <td>485</td>\n",
       "      <td>AAATAGGGAGAAATGGCGACGGAGCCTGGCTGTGGGCCCATCTTTG...</td>\n",
       "      <td>AAATAGGGAGAAATGGCGACGGAGCCTGGCTGTGGGCCCATCTTTG...</td>\n",
       "      <td>ATGCTTGAGAGACTCAAAGCCCCGTGGTCAGCTGCCCTGCAAAGAA...</td>\n",
       "      <td>TTATTCTCGATGCCCAGAGATGCTCATGCACCTGTGCACACTCACA...</td>\n",
       "      <td>2769</td>\n",
       "      <td>95</td>\n",
       "      <td>1458</td>\n",
       "      <td>1216</td>\n",
       "      <td>0.030</td>\n",
       "      <td>2.750</td>\n",
       "      <td>0.010909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7985</th>\n",
       "      <td>ENSP00000437563</td>\n",
       "      <td>ENST00000539033</td>\n",
       "      <td>ENSG00000255641</td>\n",
       "      <td>MNKQRGTFSEVSLAQDPKRQQRKPKGNKSSISGTEQEIFQVELNLQ...</td>\n",
       "      <td>240</td>\n",
       "      <td>CACACAGCTGCAGAGATGAATAAACAAAGAGGAACCTTCTCAGAAG...</td>\n",
       "      <td>CACACAGCTGCAGAG</td>\n",
       "      <td>ATGAATAAACAAAGAGGAACCTTCTCAGAAGTGAGTCTGGCCCAGG...</td>\n",
       "      <td>GCTCAAGAAATCAACACATCTTGGCCTCCCAAGTTGCTGGGATTAC...</td>\n",
       "      <td>998</td>\n",
       "      <td>15</td>\n",
       "      <td>723</td>\n",
       "      <td>260</td>\n",
       "      <td>0.102</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.927273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7986</th>\n",
       "      <td>ENSP00000483386</td>\n",
       "      <td>ENST00000619918</td>\n",
       "      <td>ENSG00000275793</td>\n",
       "      <td>MAKDSPSPLGASPKKPGCSSPAAAVLENQRRELEKLRAELEAERAG...</td>\n",
       "      <td>1639</td>\n",
       "      <td>CCAGCCACCTGACCCAAATACACCACCTGAACCCACGACCAGGTAT...</td>\n",
       "      <td>CCAGCCACCTGACCCAAATACACCACCTGAACCCACGACCAGGTAT...</td>\n",
       "      <td>ATGGCCAAGGACTCGCCCAGCCCCTTGGGCGCGTCGCCCAAGAAGC...</td>\n",
       "      <td>GCAAGCATCCTTGCCCAGGTAGTGGCCTCTGGCTGCTCACACCCTG...</td>\n",
       "      <td>6105</td>\n",
       "      <td>485</td>\n",
       "      <td>4920</td>\n",
       "      <td>700</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.010</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7987 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           protein_id          mrna_id          gene_id  \\\n",
       "0     ENSP00000341562  ENST00000341156  ENSG00000135454   \n",
       "1     ENSP00000290765  ENST00000290765  ENSG00000133433   \n",
       "2     ENSP00000256996  ENST00000256996  ENSG00000134574   \n",
       "3     ENSP00000262990  ENST00000262990  ENSG00000109445   \n",
       "4     ENSP00000371811  ENST00000382374  ENSG00000165487   \n",
       "...               ...              ...              ...   \n",
       "7982  ENSP00000387111  ENST00000410042  ENSG00000273155   \n",
       "7983  ENSP00000477037  ENST00000553909  ENSG00000259171   \n",
       "7984  ENSP00000361601  ENST00000372523  ENSG00000168612   \n",
       "7985  ENSP00000437563  ENST00000539033  ENSG00000255641   \n",
       "7986  ENSP00000483386  ENST00000619918  ENSG00000275793   \n",
       "\n",
       "                                       protein_sequence  protein_length  \\\n",
       "0     MWLGRRALCALVLLLACASLGLLYASTRDAPGLRLPLAPWAPPQSP...             533   \n",
       "1     MGLELFLDLVSQPSRAVYIFAKKNGIPLELRTVDLVKGQHKSKEFL...             244   \n",
       "2     MAPKKRPETQKTSEIVLRPRNKRSRSPLELEPEAKKLCAKGSGPSR...             427   \n",
       "3     MPKKKTGARKKAENRREREKQLRASRSTIDLAKHPCNASMECDKCQ...             320   \n",
       "4     MAAAAGSCARVAAWGGKLRRGLAVSRQAVRSPGPLAAAVAGAALAG...             434   \n",
       "...                                                 ...             ...   \n",
       "7982  MAGILRLVVQWPPGRLQTVTKGVESLICTDWIRHKFTRSRIPEKVF...             131   \n",
       "7983                    MVMGLGVLLLVFVLGLGLTPPTLAQDNSRHL              31   \n",
       "7984  MLERLKAPWSAALQRKYFDLGIWTAPISPMALTMLNGLLIKDSSPP...             485   \n",
       "7985  MNKQRGTFSEVSLAQDPKRQQRKPKGNKSSISGTEQEIFQVELNLQ...             240   \n",
       "7986  MAKDSPSPLGASPKKPGCSSPAAAVLENQRRELEKLRAELEAERAG...            1639   \n",
       "\n",
       "                                                   cdna  \\\n",
       "0     GCATTCCCCGCGCGGAGCCGAAGCAGCCGCAACGAGCCGGGAGCTG...   \n",
       "1     AGCTGCTGCCCACACCGCGCTCAGCGCCTTCACTGCCATCCCCGCT...   \n",
       "2     GGTTTGAACAAGCCCTGGGCATGTTTGGCGGGAAGTTGGCTTAGCT...   \n",
       "3     GCTGTCGTAAAAGGACGTCCGGTCCGTCTCCTAGTGTCCGGAATCG...   \n",
       "4     GCCTAGCTGCGCTTCCGCAAAGATGGCGGCGGCTGCGGGTAGCTGC...   \n",
       "...                                                 ...   \n",
       "7982  GCACTTTTCCAGCTCGAGCCCTCACGAGGCCGTGGGTACGACCGGA...   \n",
       "7983  ACACACTCACACAAGGACGCCAACCCCACCTAGATGCAAAGCAGGA...   \n",
       "7984  AAATAGGGAGAAATGGCGACGGAGCCTGGCTGTGGGCCCATCTTTG...   \n",
       "7985  CACACAGCTGCAGAGATGAATAAACAAAGAGGAACCTTCTCAGAAG...   \n",
       "7986  CCAGCCACCTGACCCAAATACACCACCTGAACCCACGACCAGGTAT...   \n",
       "\n",
       "                                                   5utr  \\\n",
       "0     GCATTCCCCGCGCGGAGCCGAAGCAGCCGCAACGAGCCGGGAGCTG...   \n",
       "1     AGCTGCTGCCCACACCGCGCTCAGCGCCTTCACTGCCATCCCCGCT...   \n",
       "2     GGTTTGAACAAGCCCTGGGCATGTTTGGCGGGAAGTTGGCTTAGCT...   \n",
       "3     GCTGTCGTAAAAGGACGTCCGGTCCGTCTCCTAGTGTCCGGAATCG...   \n",
       "4                                GCCTAGCTGCGCTTCCGCAAAG   \n",
       "...                                                 ...   \n",
       "7982  GCACTTTTCCAGCTCGAGCCCTCACGAGGCCGTGGGTACGACCGGA...   \n",
       "7983  ACACACTCACACAAGGACGCCAACCCCACCTAGATGCAAAGCAGGA...   \n",
       "7984  AAATAGGGAGAAATGGCGACGGAGCCTGGCTGTGGGCCCATCTTTG...   \n",
       "7985                                    CACACAGCTGCAGAG   \n",
       "7986  CCAGCCACCTGACCCAAATACACCACCTGAACCCACGACCAGGTAT...   \n",
       "\n",
       "                                                    cds  \\\n",
       "0     ATGTGGCTGGGCCGCCGGGCCCTGTGCGCTCTGGTCCTTCTGCTCG...   \n",
       "1     ATGGGCCTAGAGCTGTTTCTTGACCTGGTGTCCCAGCCCAGCCGCG...   \n",
       "2     ATGGCTCCCAAGAAACGCCCAGAAACCCAGAAGACCTCCGAGATTG...   \n",
       "3     ATGCCTAAAAAAAAGACTGGTGCGAGGAAGAAGGCTGAGAACCGCC...   \n",
       "4     ATGGCGGCGGCTGCGGGTAGCTGCGCGCGGGTGGCGGCCTGGGGCG...   \n",
       "...                                                 ...   \n",
       "7982  ATGGCTGGGATTTTGCGCTTAGTAGTTCAATGGCCCCCAGGCAGAC...   \n",
       "7983  ATGGTGATGGGCCTGGGCGTTTTGTTGTTGGTCTTCGTGCTGGGTC...   \n",
       "7984  ATGCTTGAGAGACTCAAAGCCCCGTGGTCAGCTGCCCTGCAAAGAA...   \n",
       "7985  ATGAATAAACAAAGAGGAACCTTCTCAGAAGTGAGTCTGGCCCAGG...   \n",
       "7986  ATGGCCAAGGACTCGCCCAGCCCCTTGGGCGCGTCGCCCAAGAAGC...   \n",
       "\n",
       "                                                   3utr  cdna_length  \\\n",
       "0     TGGCCCGCTGGGGATTTCTGACTGTCAGGCTGGGCCTGCCTCCTTG...         5368   \n",
       "1     AGGGTCTGGGATGGGGGCCAGGAGATTAGCAACAAGGATTCATTCT...         1108   \n",
       "2     GAGACACTAAAGAAGGTGTGGGCCAGACAAGGCCTTGGAGCCCACA...         1815   \n",
       "3     GGGAGCTGCTCTGGTGGCCGTGTGTGAGAGGAGCAGGAGTGAGTGT...         1891   \n",
       "4     TAAAAGATATAATAGTATGGCAATTATATTGTTCCAAATGTCAAAA...         1885   \n",
       "...                                                 ...          ...   \n",
       "7982  TGTAATTCTAAATGCATTTTTCTTTTTTCGTTAATGTTCTGAAAAA...          669   \n",
       "7983  GATACTGATGGCTCTGCAGAGGACCCATTCATTGCTTCTGCTTTTG...         1498   \n",
       "7984  TTATTCTCGATGCCCAGAGATGCTCATGCACCTGTGCACACTCACA...         2769   \n",
       "7985  GCTCAAGAAATCAACACATCTTGGCCTCCCAAGTTGCTGGGATTAC...          998   \n",
       "7986  GCAAGCATCCTTGCCCAGGTAGTGGCCTCTGGCTGCTCACACCCTG...         6105   \n",
       "\n",
       "      5utr_length  cds_length  3utr_length  protein_abundance  median_TPM  \\\n",
       "0             431        1602         3335              0.068       0.100   \n",
       "1              64         735          309             57.000      11.385   \n",
       "2             163        1284          368              2.510       3.740   \n",
       "3             220         963          708              2.500      13.445   \n",
       "4              22        1305          558              2.500       8.060   \n",
       "...           ...         ...          ...                ...         ...   \n",
       "7982          207         396           66              1.010       0.155   \n",
       "7983          190          96         1212              0.591      24.155   \n",
       "7984           95        1458         1216              0.030       2.750   \n",
       "7985           15         723          260              0.102       0.110   \n",
       "7986          485        4920          700              0.010       0.010   \n",
       "\n",
       "      protein_per_transcript  \n",
       "0                   0.680000  \n",
       "1                   5.006588  \n",
       "2                   0.671123  \n",
       "3                   0.185943  \n",
       "4                   0.310174  \n",
       "...                      ...  \n",
       "7982                6.516129  \n",
       "7983                0.024467  \n",
       "7984                0.010909  \n",
       "7985                0.927273  \n",
       "7986                1.000000  \n",
       "\n",
       "[7987 rows x 16 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df1 = pd.read_csv(\"/Users/alvinthosatria/Documents/FYP/datasets/cleaned_liver_data_cdna.csv\")\n",
    "\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/numpy/core/include\n",
      "/Library/Frameworks/Python.framework/Versions/3.8/include/python3.8\n"
     ]
    }
   ],
   "source": [
    "import numpy\n",
    "import sysconfig\n",
    "\n",
    "print(numpy.get_include())\n",
    "print(sysconfig.get_path('include'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pip in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (25.0.1)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: cython==0.29 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (0.29)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: numpy in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (1.24.4)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: biopython==1.78 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (1.78)\n",
      "Requirement already satisfied: numpy in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from biopython==1.78) (1.24.4)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: editdistance in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (0.8.1)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: torch==1.7.1 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (1.7.1)\n",
      "Requirement already satisfied: torchvision==0.8.2 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (0.8.2)\n",
      "Requirement already satisfied: torchaudio==0.7.2 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (0.7.2)\n",
      "Requirement already satisfied: typing-extensions in /Users/alvinthosatria/Library/Python/3.8/lib/python/site-packages (from torch==1.7.1) (4.12.2)\n",
      "Requirement already satisfied: numpy in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch==1.7.1) (1.24.4)\n",
      "Requirement already satisfied: pillow>=4.1.1 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torchvision==0.8.2) (10.4.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: torch_geometric==1.7.0 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (1.7.0)\n",
      "Requirement already satisfied: numpy in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (1.24.4)\n",
      "Requirement already satisfied: tqdm in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (4.66.6)\n",
      "Requirement already satisfied: scipy in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (1.10.1)\n",
      "Requirement already satisfied: networkx in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (3.1)\n",
      "Requirement already satisfied: python-louvain in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (0.16)\n",
      "Requirement already satisfied: scikit-learn in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (1.3.2)\n",
      "Requirement already satisfied: numba in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (0.58.1)\n",
      "Requirement already satisfied: requests in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (2.32.3)\n",
      "Requirement already satisfied: pandas in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (2.0.3)\n",
      "Requirement already satisfied: rdflib in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (7.1.3)\n",
      "Requirement already satisfied: h5py in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (3.11.0)\n",
      "Requirement already satisfied: googledrivedownloader in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (1.1.0)\n",
      "Requirement already satisfied: ase in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (3.23.0)\n",
      "Requirement already satisfied: jinja2 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch_geometric==1.7.0) (3.1.4)\n",
      "Requirement already satisfied: matplotlib>=3.3.4 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from ase->torch_geometric==1.7.0) (3.7.5)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from jinja2->torch_geometric==1.7.0) (2.1.5)\n",
      "Requirement already satisfied: llvmlite<0.42,>=0.41.0dev0 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from numba->torch_geometric==1.7.0) (0.41.1)\n",
      "Requirement already satisfied: importlib-metadata in /Users/alvinthosatria/Library/Python/3.8/lib/python/site-packages (from numba->torch_geometric==1.7.0) (8.5.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Users/alvinthosatria/Library/Python/3.8/lib/python/site-packages (from pandas->torch_geometric==1.7.0) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from pandas->torch_geometric==1.7.0) (2025.1)\n",
      "Requirement already satisfied: tzdata>=2022.1 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from pandas->torch_geometric==1.7.0) (2025.1)\n",
      "Requirement already satisfied: isodate<1.0.0,>=0.7.2 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from rdflib->torch_geometric==1.7.0) (0.7.2)\n",
      "Requirement already satisfied: pyparsing<4,>=2.1.0 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from rdflib->torch_geometric==1.7.0) (3.1.4)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from requests->torch_geometric==1.7.0) (3.4.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from requests->torch_geometric==1.7.0) (3.10)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from requests->torch_geometric==1.7.0) (2.2.3)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from requests->torch_geometric==1.7.0) (2024.8.30)\n",
      "Requirement already satisfied: joblib>=1.1.1 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from scikit-learn->torch_geometric==1.7.0) (1.4.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from scikit-learn->torch_geometric==1.7.0) (3.5.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from matplotlib>=3.3.4->ase->torch_geometric==1.7.0) (1.1.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from matplotlib>=3.3.4->ase->torch_geometric==1.7.0) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from matplotlib>=3.3.4->ase->torch_geometric==1.7.0) (4.56.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from matplotlib>=3.3.4->ase->torch_geometric==1.7.0) (1.4.7)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/alvinthosatria/Library/Python/3.8/lib/python/site-packages (from matplotlib>=3.3.4->ase->torch_geometric==1.7.0) (24.1)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from matplotlib>=3.3.4->ase->torch_geometric==1.7.0) (10.4.0)\n",
      "Requirement already satisfied: importlib-resources>=3.2.0 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from matplotlib>=3.3.4->ase->torch_geometric==1.7.0) (6.4.5)\n",
      "Requirement already satisfied: six>=1.5 in /Users/alvinthosatria/Library/Python/3.8/lib/python/site-packages (from python-dateutil>=2.8.2->pandas->torch_geometric==1.7.0) (1.16.0)\n",
      "Requirement already satisfied: zipp>=3.20 in /Users/alvinthosatria/Library/Python/3.8/lib/python/site-packages (from importlib-metadata->numba->torch_geometric==1.7.0) (3.20.2)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: pip in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (25.0.1)\n",
      "Requirement already satisfied: setuptools in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (75.3.2)\n",
      "Requirement already satisfied: wheel in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (0.45.1)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: ninja in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (1.11.1.3)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: cmake in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (3.31.4)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install --upgrade pip\n",
    "%pip install cython==0.29\n",
    "%pip install numpy\n",
    "%pip install biopython==1.78\n",
    "%pip install editdistance\n",
    "%pip install torch==1.7.1 torchvision==0.8.2 torchaudio==0.7.2\n",
    "%pip install torch_geometric==1.7.0\n",
    "%pip install --upgrade pip setuptools wheel\n",
    "%pip install ninja\n",
    "%pip install cmake"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: torch-sparse==0.6.9 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (0.6.9)\n",
      "Requirement already satisfied: scipy in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from torch-sparse==0.6.9) (1.10.1)\n",
      "Requirement already satisfied: numpy<1.27.0,>=1.19.5 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (from scipy->torch-sparse==0.6.9) (1.24.4)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: torch-scatter==2.0.7 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (2.0.7)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install torch-sparse==0.6.9\n",
    "%pip install torch-scatter==2.0.7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3.8.3 (v3.8.3:6f8c8320e9, May 13 2020, 16:29:34) \\n[Clang 6.0 (clang-600.0.57)]'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "sys.version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fatal: destination path 'GCNFrame' already exists and is not an empty directory.\n",
      "/Users/alvinthosatria/Documents/FYP/datasets/Sequence_Embeddings/mRNA_Sequence_Embeddings/GP-GCN/GCNFrame/GCNFrame\n",
      "/Users/alvinthosatria/Documents/FYP/datasets/Sequence_Embeddings/mRNA_Sequence_Embeddings/GP-GCN/GCNFrame\n"
     ]
    }
   ],
   "source": [
    "!git clone https://github.com/deepomicslab/GCNFrame.git\n",
    "%cd GCNFrame/GCNFrame\n",
    "import sys\n",
    "!{sys.executable} setup.py build_ext --inplace\n",
    "%cd ../"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pre-train Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Encoding sequences...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GraphDatasetInMem(1551)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import sys\n",
    "# sys.path.append(\"/Users/alvinthosatria/Documents/FYP/datasets/Sequence_Embeddings/mRNA_Sequence_Embeddings/GP-GCN/GCNFrame\")\n",
    "from GCNFrame import Biodata, GCNmodel\n",
    "import torch\n",
    "\n",
    "data = Biodata(\n",
    "    fasta_file=\"/Users/alvinthosatria/Documents/FYP/datasets/Sequence_Embeddings/mRNA_Sequence_Embeddings/GP-GCN/example_data/nature_2017.fasta\",\n",
    "    label_file=\"/Users/alvinthosatria/Documents/FYP/datasets/Sequence_Embeddings/mRNA_Sequence_Embeddings/GP-GCN/example_data/lifestyle_label.txt\",\n",
    "    feature_file=\"/Users/alvinthosatria/Documents/FYP/datasets/Sequence_Embeddings/mRNA_Sequence_Embeddings/GP-GCN/example_data/CDD_protein_feature.txt\"\n",
    ")\n",
    "\n",
    "dataset = data.encode(thread=20)\n",
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 0| Loss: 0.6426| Train accuracy: 0.7826| Validation accuracy: 0.8710\n",
      "Epoch 1| Loss: 0.5558| Train accuracy: 0.8526| Validation accuracy: 0.7935\n",
      "Epoch 2| Loss: 0.5117| Train accuracy: 0.8165| Validation accuracy: 0.8516\n",
      "Epoch 3| Loss: 0.4632| Train accuracy: 0.8658| Validation accuracy: 0.8387\n",
      "Epoch 4| Loss: 0.4606| Train accuracy: 0.8643| Validation accuracy: 0.8581\n",
      "Epoch 5| Loss: 0.4297| Train accuracy: 0.8942| Validation accuracy: 0.8516\n",
      "Epoch 6| Loss: 0.4239| Train accuracy: 0.8958| Validation accuracy: 0.8323\n",
      "Epoch 7| Loss: 0.4249| Train accuracy: 0.8925| Validation accuracy: 0.8516\n",
      "Epoch 8| Loss: 0.4230| Train accuracy: 0.8992| Validation accuracy: 0.8581\n",
      "Epoch 9| Loss: 0.4024| Train accuracy: 0.9201| Validation accuracy: 0.8516\n",
      "Epoch 10| Loss: 0.4057| Train accuracy: 0.9164| Validation accuracy: 0.8839\n",
      "Epoch 11| Loss: 0.3976| Train accuracy: 0.9256| Validation accuracy: 0.8774\n",
      "Epoch 12| Loss: 0.3961| Train accuracy: 0.9281| Validation accuracy: 0.8839\n",
      "Epoch 13| Loss: 0.3802| Train accuracy: 0.9406| Validation accuracy: 0.8258\n",
      "Epoch 14| Loss: 0.3846| Train accuracy: 0.9354| Validation accuracy: 0.8839\n",
      "Epoch 15| Loss: 0.3643| Train accuracy: 0.9567| Validation accuracy: 0.9032\n",
      "Epoch 16| Loss: 0.3696| Train accuracy: 0.9459| Validation accuracy: 0.8968\n",
      "Epoch 17| Loss: 0.3688| Train accuracy: 0.9530| Validation accuracy: 0.8645\n",
      "Epoch 18| Loss: 0.3641| Train accuracy: 0.9503| Validation accuracy: 0.9032\n",
      "Epoch 19| Loss: 0.3709| Train accuracy: 0.9434| Validation accuracy: 0.9097\n",
      "Encoding sequences...\n"
     ]
    }
   ],
   "source": [
    "device = torch.device(\"cpu\")\n",
    "\n",
    "model = GCNmodel.model(label_num=2, other_feature_dim=206).to(device)\n",
    "GCNmodel.train(dataset, model, weighted_sampling=True)\n",
    "GCNmodel.test(\n",
    "    model_name=\"GCN_model.pt\",\n",
    "    fasta_file=\"/Users/alvinthosatria/Documents/FYP/datasets/Sequence_Embeddings/mRNA_Sequence_Embeddings/GP-GCN/example_data/CDD_protein_feature.txt\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use own data to extract embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping layer d2.weight due to shape mismatch or incompatibility.\n",
      "Skipping layer d2.bias due to shape mismatch or incompatibility.\n",
      "Skipping layer d3.weight due to shape mismatch or incompatibility.\n",
      "Skipping layer d3.bias due to shape mismatch or incompatibility.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "model(\n",
       "  (pnode_d): Linear(in_features=12288, out_features=12288, bias=True)\n",
       "  (fnode_d): Linear(in_features=64, out_features=192, bias=True)\n",
       "  (gconvs_1): ModuleList(\n",
       "    (0): SAGEConv((3, 3), 128)\n",
       "    (1): SAGEConv((128, 128), 128)\n",
       "  )\n",
       "  (gconvs_2): ModuleList(\n",
       "    (0): SAGEConv((128, 3), 128)\n",
       "    (1): SAGEConv((128, 128), 128)\n",
       "  )\n",
       "  (lns): ModuleList(\n",
       "    (0): LayerNorm((128,), eps=1e-05, elementwise_affine=True)\n",
       "  )\n",
       "  (convs): ModuleList(\n",
       "    (0): Conv1d(128, 64, kernel_size=(8,), stride=(1,))\n",
       "    (1): Conv1d(64, 64, kernel_size=(8,), stride=(1,))\n",
       "    (2): Conv1d(64, 64, kernel_size=(8,), stride=(1,))\n",
       "  )\n",
       "  (d1): Linear(in_features=260800, out_features=100, bias=True)\n",
       "  (d2): Linear(in_features=100, out_features=2, bias=True)\n",
       ")"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import torch\n",
    "\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "model = GCNmodel.model(label_num=2, other_feature_dim=None).to(device)\n",
    "model_name = \"/Users/alvinthosatria/Documents/FYP/datasets/Sequence_Embeddings/mRNA_Sequence_Embeddings/GP-GCN/GCNFrame/GCN_model.pt\"\n",
    "saved_model = torch.load(model_name, map_location=device)\n",
    "saved_state_dict = saved_model.state_dict()\n",
    "new_state_dict = model.state_dict()\n",
    "\n",
    "for name, param in saved_state_dict.items():\n",
    "    if name in new_state_dict and new_state_dict[name].shape == param.shape:\n",
    "        new_state_dict[name].copy_(param)\n",
    "    else:\n",
    "        print(f\"Skipping layer {name} due to shape mismatch or incompatibility.\")\n",
    "\n",
    "model.load_state_dict(new_state_dict, strict=False)\n",
    "\n",
    "model.eval()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "fasta_file = \"/Users/alvinthosatria/Documents/FYP/datasets/Sequence_Embeddings/mRNA_Sequence_Embeddings/GP-GCN/cdna_data/liver_cdna.fasta\"\n",
    "fasta_sequences = list(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "total_sequences = len(fasta_sequences)\n",
    "batch_size = 1551\n",
    "batches = [fasta_sequences[i:i + batch_size] for i in range(0, total_sequences, batch_size)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing batch 1/6 with 1551 sequences\n",
      "Encoding sequences...\n",
      "Number of sequences in batch_dataset: 1551\n",
      "Shape of batch_embeddings: (1551, 100)\n",
      "\n",
      "Processing batch 2/6 with 1551 sequences\n",
      "Encoding sequences...\n",
      "Number of sequences in batch_dataset: 1551\n",
      "Shape of batch_embeddings: (1551, 100)\n",
      "\n",
      "Processing batch 3/6 with 1551 sequences\n",
      "Encoding sequences...\n",
      "Number of sequences in batch_dataset: 1551\n",
      "Shape of batch_embeddings: (1551, 100)\n",
      "\n",
      "Processing batch 4/6 with 1551 sequences\n",
      "Encoding sequences...\n",
      "Number of sequences in batch_dataset: 1551\n",
      "Shape of batch_embeddings: (1551, 100)\n",
      "\n",
      "Processing batch 5/6 with 1551 sequences\n",
      "Encoding sequences...\n",
      "Number of sequences in batch_dataset: 1551\n",
      "Shape of batch_embeddings: (1551, 100)\n",
      "\n",
      "Processing batch 6/6 with 232 sequences\n",
      "Encoding sequences...\n",
      "Number of sequences in batch_dataset: 232\n",
      "Shape of batch_embeddings: (232, 100)\n",
      "Final shape of all_embeddings: (7987, 100)\n",
      "        protein_id          mrna_id     emb_0     emb_1     emb_2     emb_3  \\\n",
      "0  ENSP00000341562  ENST00000341156 -3.652156 -4.719733 -4.365098 -4.977760   \n",
      "1  ENSP00000290765  ENST00000290765 -3.380323 -4.438233 -4.082952 -4.748327   \n",
      "2  ENSP00000256996  ENST00000256996 -3.477070 -4.537240 -4.144571 -4.833215   \n",
      "3  ENSP00000262990  ENST00000262990 -3.312737 -4.642695 -4.138601 -4.664500   \n",
      "4  ENSP00000371811  ENST00000382374 -3.311863 -4.548461 -4.043988 -4.392979   \n",
      "\n",
      "      emb_4     emb_5     emb_6     emb_7  ...    emb_90    emb_91    emb_92  \\\n",
      "0 -6.878171 -5.098008 -0.985231 -6.352939  ...  8.402204 -4.746917 -3.176349   \n",
      "1 -6.462082 -4.827402  4.282731 -5.913006  ...  8.692030 -4.506042 -3.105223   \n",
      "2 -6.583676 -4.939151  3.533929 -6.037057  ...  9.281156 -4.633046 -3.134871   \n",
      "3 -6.500449 -5.128020  8.008026 -5.895562  ...  8.624372 -4.572633 -3.054109   \n",
      "4 -6.373192 -5.062601  8.503234 -5.803919  ...  8.666829 -4.391415 -2.872576   \n",
      "\n",
      "     emb_93    emb_94    emb_95    emb_96    emb_97     emb_98    emb_99  \n",
      "0 -1.070461 -7.316020 -1.820091 -2.311639 -2.233166   9.429097 -6.576334  \n",
      "1  4.381436 -6.897438 -2.132094  3.515608 -2.076446   9.707837 -6.246079  \n",
      "2  3.499077 -7.039967 -2.164059  2.530649 -2.093006   9.875761 -6.357322  \n",
      "3  8.472735 -6.971133 -2.716433  7.566904 -2.274932  10.166460 -6.333469  \n",
      "4  8.967222 -6.773387 -2.567684  9.243492 -2.304719  10.998148 -6.195429  \n",
      "\n",
      "[5 rows x 102 columns]\n",
      "Embeddings DataFrame saved to liver_cdna_embeddings.csv\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "from torch_geometric.data import DataLoader\n",
    "from GCNFrame import Biodata\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "\n",
    "all_embeddings = []\n",
    "all_protein_ids = []\n",
    "all_mrna_ids = []\n",
    "\n",
    "for batch_idx, batch_sequences in enumerate(batches):\n",
    "    print(f\"\\nProcessing batch {batch_idx + 1}/{len(batches)} with {len(batch_sequences)} sequences\")\n",
    "    \n",
    "    batch_fasta_file = f\"batch_{batch_idx}.fasta\"\n",
    "    with open(batch_fasta_file, \"w\") as f:\n",
    "        SeqIO.write(batch_sequences, f, \"fasta\")\n",
    "    \n",
    "    data = Biodata(fasta_file=batch_fasta_file)\n",
    "    batch_dataset = data.encode(thread=20, save_dataset=False)\n",
    "    print(f\"Number of sequences in batch_dataset: {len(batch_dataset)}\")\n",
    "    \n",
    "    if len(batch_dataset) != len(batch_sequences):\n",
    "        raise ValueError(f\"Batch {batch_idx + 1} has {len(batch_sequences)} sequences, but batch_dataset has {len(batch_dataset)} sequences.\")\n",
    "    \n",
    "    batch_loader = DataLoader(batch_dataset, batch_size=32, shuffle=False, follow_batch=['x_src', 'x_dst'])\n",
    "    \n",
    "    batch_embeddings = []\n",
    "    with torch.no_grad():\n",
    "        for batch in batch_loader:\n",
    "            batch = batch.to(device)\n",
    "            out, embeddings = model(batch)\n",
    "            embeddings = embeddings.cpu().numpy()\n",
    "            batch_embeddings.append(embeddings)\n",
    "    \n",
    "    batch_embeddings = np.concatenate(batch_embeddings, axis=0)\n",
    "    print(f\"Shape of batch_embeddings: {batch_embeddings.shape}\")\n",
    "    \n",
    "    all_embeddings.append(batch_embeddings)\n",
    "    \n",
    "    for seq_record in batch_sequences:\n",
    "        header_parts = seq_record.id.split(\"|\")\n",
    "        if len(header_parts) >= 2:\n",
    "            all_protein_ids.append(header_parts[0])\n",
    "            all_mrna_ids.append(header_parts[1])\n",
    "        else:\n",
    "            raise ValueError(f\"FASTA header {seq_record.id} does not contain both protein_id and mrna_id\")\n",
    "    \n",
    "    os.remove(batch_fasta_file)\n",
    "\n",
    "all_embeddings = np.concatenate(all_embeddings, axis=0)\n",
    "print(f\"Final shape of all_embeddings: {all_embeddings.shape}\")  # Should be (7987, 100)\n",
    "\n",
    "if all_embeddings.shape[0] != total_sequences:\n",
    "    raise ValueError(f\"Expected {total_sequences} embeddings, but got {all_embeddings.shape[0]}.\")\n",
    "\n",
    "embedding_dim = all_embeddings.shape[1]\n",
    "embedding_columns = [f\"emb_{i}\" for i in range(embedding_dim)]\n",
    "\n",
    "df = pd.DataFrame(\n",
    "    data=np.column_stack([all_protein_ids, all_mrna_ids, all_embeddings]),\n",
    "    columns=['protein_id', 'mrna_id'] + embedding_columns\n",
    ")\n",
    "\n",
    "df[embedding_columns] = df[embedding_columns].apply(pd.to_numeric)\n",
    "\n",
    "print(df.head())\n",
    "\n",
    "df.to_csv(\"liver_cdna_embeddings.csv\", index=False)\n",
    "print(\"Embeddings DataFrame saved to liver_cdna_embeddings.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>emb_0</th>\n",
       "      <th>emb_1</th>\n",
       "      <th>emb_2</th>\n",
       "      <th>emb_3</th>\n",
       "      <th>emb_4</th>\n",
       "      <th>emb_5</th>\n",
       "      <th>emb_6</th>\n",
       "      <th>emb_7</th>\n",
       "      <th>...</th>\n",
       "      <th>emb_90</th>\n",
       "      <th>emb_91</th>\n",
       "      <th>emb_92</th>\n",
       "      <th>emb_93</th>\n",
       "      <th>emb_94</th>\n",
       "      <th>emb_95</th>\n",
       "      <th>emb_96</th>\n",
       "      <th>emb_97</th>\n",
       "      <th>emb_98</th>\n",
       "      <th>emb_99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000341562</td>\n",
       "      <td>ENST00000341156</td>\n",
       "      <td>-3.652156</td>\n",
       "      <td>-4.719733</td>\n",
       "      <td>-4.365098</td>\n",
       "      <td>-4.977760</td>\n",
       "      <td>-6.878171</td>\n",
       "      <td>-5.098008</td>\n",
       "      <td>-0.985231</td>\n",
       "      <td>-6.352939</td>\n",
       "      <td>...</td>\n",
       "      <td>8.402204</td>\n",
       "      <td>-4.746917</td>\n",
       "      <td>-3.176349</td>\n",
       "      <td>-1.070461</td>\n",
       "      <td>-7.316020</td>\n",
       "      <td>-1.820091</td>\n",
       "      <td>-2.311639</td>\n",
       "      <td>-2.233166</td>\n",
       "      <td>9.429097</td>\n",
       "      <td>-6.576334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000290765</td>\n",
       "      <td>ENST00000290765</td>\n",
       "      <td>-3.380323</td>\n",
       "      <td>-4.438233</td>\n",
       "      <td>-4.082952</td>\n",
       "      <td>-4.748327</td>\n",
       "      <td>-6.462082</td>\n",
       "      <td>-4.827402</td>\n",
       "      <td>4.282731</td>\n",
       "      <td>-5.913006</td>\n",
       "      <td>...</td>\n",
       "      <td>8.692030</td>\n",
       "      <td>-4.506042</td>\n",
       "      <td>-3.105223</td>\n",
       "      <td>4.381436</td>\n",
       "      <td>-6.897438</td>\n",
       "      <td>-2.132094</td>\n",
       "      <td>3.515608</td>\n",
       "      <td>-2.076446</td>\n",
       "      <td>9.707837</td>\n",
       "      <td>-6.246079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000256996</td>\n",
       "      <td>ENST00000256996</td>\n",
       "      <td>-3.477070</td>\n",
       "      <td>-4.537240</td>\n",
       "      <td>-4.144571</td>\n",
       "      <td>-4.833215</td>\n",
       "      <td>-6.583676</td>\n",
       "      <td>-4.939151</td>\n",
       "      <td>3.533929</td>\n",
       "      <td>-6.037057</td>\n",
       "      <td>...</td>\n",
       "      <td>9.281156</td>\n",
       "      <td>-4.633046</td>\n",
       "      <td>-3.134871</td>\n",
       "      <td>3.499077</td>\n",
       "      <td>-7.039967</td>\n",
       "      <td>-2.164059</td>\n",
       "      <td>2.530649</td>\n",
       "      <td>-2.093006</td>\n",
       "      <td>9.875761</td>\n",
       "      <td>-6.357322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000262990</td>\n",
       "      <td>ENST00000262990</td>\n",
       "      <td>-3.312737</td>\n",
       "      <td>-4.642695</td>\n",
       "      <td>-4.138601</td>\n",
       "      <td>-4.664500</td>\n",
       "      <td>-6.500449</td>\n",
       "      <td>-5.128020</td>\n",
       "      <td>8.008026</td>\n",
       "      <td>-5.895562</td>\n",
       "      <td>...</td>\n",
       "      <td>8.624372</td>\n",
       "      <td>-4.572633</td>\n",
       "      <td>-3.054109</td>\n",
       "      <td>8.472735</td>\n",
       "      <td>-6.971133</td>\n",
       "      <td>-2.716433</td>\n",
       "      <td>7.566904</td>\n",
       "      <td>-2.274932</td>\n",
       "      <td>10.166460</td>\n",
       "      <td>-6.333469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000371811</td>\n",
       "      <td>ENST00000382374</td>\n",
       "      <td>-3.311863</td>\n",
       "      <td>-4.548461</td>\n",
       "      <td>-4.043988</td>\n",
       "      <td>-4.392979</td>\n",
       "      <td>-6.373192</td>\n",
       "      <td>-5.062601</td>\n",
       "      <td>8.503234</td>\n",
       "      <td>-5.803919</td>\n",
       "      <td>...</td>\n",
       "      <td>8.666829</td>\n",
       "      <td>-4.391415</td>\n",
       "      <td>-2.872576</td>\n",
       "      <td>8.967222</td>\n",
       "      <td>-6.773387</td>\n",
       "      <td>-2.567684</td>\n",
       "      <td>9.243492</td>\n",
       "      <td>-2.304719</td>\n",
       "      <td>10.998148</td>\n",
       "      <td>-6.195429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7982</th>\n",
       "      <td>ENSP00000387111</td>\n",
       "      <td>ENST00000410042</td>\n",
       "      <td>-3.540463</td>\n",
       "      <td>-4.676218</td>\n",
       "      <td>-4.206406</td>\n",
       "      <td>-4.563828</td>\n",
       "      <td>-6.625822</td>\n",
       "      <td>-5.097000</td>\n",
       "      <td>1.477638</td>\n",
       "      <td>-6.119875</td>\n",
       "      <td>...</td>\n",
       "      <td>8.860622</td>\n",
       "      <td>-4.505900</td>\n",
       "      <td>-2.830032</td>\n",
       "      <td>1.688760</td>\n",
       "      <td>-7.050861</td>\n",
       "      <td>-1.896498</td>\n",
       "      <td>2.198244</td>\n",
       "      <td>-2.298310</td>\n",
       "      <td>11.843821</td>\n",
       "      <td>-6.402756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7983</th>\n",
       "      <td>ENSP00000477037</td>\n",
       "      <td>ENST00000553909</td>\n",
       "      <td>-3.381349</td>\n",
       "      <td>-4.630548</td>\n",
       "      <td>-4.195150</td>\n",
       "      <td>-4.833974</td>\n",
       "      <td>-6.639976</td>\n",
       "      <td>-5.204910</td>\n",
       "      <td>5.986599</td>\n",
       "      <td>-6.016151</td>\n",
       "      <td>...</td>\n",
       "      <td>7.996094</td>\n",
       "      <td>-4.624668</td>\n",
       "      <td>-3.126696</td>\n",
       "      <td>6.551436</td>\n",
       "      <td>-7.105997</td>\n",
       "      <td>-2.703525</td>\n",
       "      <td>5.318639</td>\n",
       "      <td>-2.245374</td>\n",
       "      <td>9.741588</td>\n",
       "      <td>-6.433317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7984</th>\n",
       "      <td>ENSP00000361601</td>\n",
       "      <td>ENST00000372523</td>\n",
       "      <td>-3.754153</td>\n",
       "      <td>-4.919304</td>\n",
       "      <td>-4.479662</td>\n",
       "      <td>-5.100664</td>\n",
       "      <td>-7.076601</td>\n",
       "      <td>-5.318255</td>\n",
       "      <td>-2.107743</td>\n",
       "      <td>-6.482817</td>\n",
       "      <td>...</td>\n",
       "      <td>8.662652</td>\n",
       "      <td>-4.802837</td>\n",
       "      <td>-3.177442</td>\n",
       "      <td>-2.163506</td>\n",
       "      <td>-7.503860</td>\n",
       "      <td>-1.791874</td>\n",
       "      <td>-3.342985</td>\n",
       "      <td>-2.279790</td>\n",
       "      <td>10.058605</td>\n",
       "      <td>-6.755480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7985</th>\n",
       "      <td>ENSP00000437563</td>\n",
       "      <td>ENST00000539033</td>\n",
       "      <td>-3.272629</td>\n",
       "      <td>-4.512104</td>\n",
       "      <td>-4.030314</td>\n",
       "      <td>-4.526674</td>\n",
       "      <td>-6.402953</td>\n",
       "      <td>-5.004708</td>\n",
       "      <td>6.798046</td>\n",
       "      <td>-5.813955</td>\n",
       "      <td>...</td>\n",
       "      <td>8.063452</td>\n",
       "      <td>-4.437620</td>\n",
       "      <td>-2.924208</td>\n",
       "      <td>7.474921</td>\n",
       "      <td>-6.809915</td>\n",
       "      <td>-2.625396</td>\n",
       "      <td>7.132857</td>\n",
       "      <td>-2.230507</td>\n",
       "      <td>10.495076</td>\n",
       "      <td>-6.194475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7986</th>\n",
       "      <td>ENSP00000483386</td>\n",
       "      <td>ENST00000619918</td>\n",
       "      <td>-3.574793</td>\n",
       "      <td>-4.495584</td>\n",
       "      <td>-4.235926</td>\n",
       "      <td>-4.776767</td>\n",
       "      <td>-6.551501</td>\n",
       "      <td>-4.869722</td>\n",
       "      <td>-1.581470</td>\n",
       "      <td>-6.028342</td>\n",
       "      <td>...</td>\n",
       "      <td>8.364728</td>\n",
       "      <td>-4.524833</td>\n",
       "      <td>-2.978835</td>\n",
       "      <td>-1.967394</td>\n",
       "      <td>-6.929486</td>\n",
       "      <td>-1.390077</td>\n",
       "      <td>-2.642559</td>\n",
       "      <td>-2.139292</td>\n",
       "      <td>9.799885</td>\n",
       "      <td>-6.237199</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7987 rows × 102 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           protein_id          mrna_id     emb_0     emb_1     emb_2  \\\n",
       "0     ENSP00000341562  ENST00000341156 -3.652156 -4.719733 -4.365098   \n",
       "1     ENSP00000290765  ENST00000290765 -3.380323 -4.438233 -4.082952   \n",
       "2     ENSP00000256996  ENST00000256996 -3.477070 -4.537240 -4.144571   \n",
       "3     ENSP00000262990  ENST00000262990 -3.312737 -4.642695 -4.138601   \n",
       "4     ENSP00000371811  ENST00000382374 -3.311863 -4.548461 -4.043988   \n",
       "...               ...              ...       ...       ...       ...   \n",
       "7982  ENSP00000387111  ENST00000410042 -3.540463 -4.676218 -4.206406   \n",
       "7983  ENSP00000477037  ENST00000553909 -3.381349 -4.630548 -4.195150   \n",
       "7984  ENSP00000361601  ENST00000372523 -3.754153 -4.919304 -4.479662   \n",
       "7985  ENSP00000437563  ENST00000539033 -3.272629 -4.512104 -4.030314   \n",
       "7986  ENSP00000483386  ENST00000619918 -3.574793 -4.495584 -4.235926   \n",
       "\n",
       "         emb_3     emb_4     emb_5     emb_6     emb_7  ...    emb_90  \\\n",
       "0    -4.977760 -6.878171 -5.098008 -0.985231 -6.352939  ...  8.402204   \n",
       "1    -4.748327 -6.462082 -4.827402  4.282731 -5.913006  ...  8.692030   \n",
       "2    -4.833215 -6.583676 -4.939151  3.533929 -6.037057  ...  9.281156   \n",
       "3    -4.664500 -6.500449 -5.128020  8.008026 -5.895562  ...  8.624372   \n",
       "4    -4.392979 -6.373192 -5.062601  8.503234 -5.803919  ...  8.666829   \n",
       "...        ...       ...       ...       ...       ...  ...       ...   \n",
       "7982 -4.563828 -6.625822 -5.097000  1.477638 -6.119875  ...  8.860622   \n",
       "7983 -4.833974 -6.639976 -5.204910  5.986599 -6.016151  ...  7.996094   \n",
       "7984 -5.100664 -7.076601 -5.318255 -2.107743 -6.482817  ...  8.662652   \n",
       "7985 -4.526674 -6.402953 -5.004708  6.798046 -5.813955  ...  8.063452   \n",
       "7986 -4.776767 -6.551501 -4.869722 -1.581470 -6.028342  ...  8.364728   \n",
       "\n",
       "        emb_91    emb_92    emb_93    emb_94    emb_95    emb_96    emb_97  \\\n",
       "0    -4.746917 -3.176349 -1.070461 -7.316020 -1.820091 -2.311639 -2.233166   \n",
       "1    -4.506042 -3.105223  4.381436 -6.897438 -2.132094  3.515608 -2.076446   \n",
       "2    -4.633046 -3.134871  3.499077 -7.039967 -2.164059  2.530649 -2.093006   \n",
       "3    -4.572633 -3.054109  8.472735 -6.971133 -2.716433  7.566904 -2.274932   \n",
       "4    -4.391415 -2.872576  8.967222 -6.773387 -2.567684  9.243492 -2.304719   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "7982 -4.505900 -2.830032  1.688760 -7.050861 -1.896498  2.198244 -2.298310   \n",
       "7983 -4.624668 -3.126696  6.551436 -7.105997 -2.703525  5.318639 -2.245374   \n",
       "7984 -4.802837 -3.177442 -2.163506 -7.503860 -1.791874 -3.342985 -2.279790   \n",
       "7985 -4.437620 -2.924208  7.474921 -6.809915 -2.625396  7.132857 -2.230507   \n",
       "7986 -4.524833 -2.978835 -1.967394 -6.929486 -1.390077 -2.642559 -2.139292   \n",
       "\n",
       "         emb_98    emb_99  \n",
       "0      9.429097 -6.576334  \n",
       "1      9.707837 -6.246079  \n",
       "2      9.875761 -6.357322  \n",
       "3     10.166460 -6.333469  \n",
       "4     10.998148 -6.195429  \n",
       "...         ...       ...  \n",
       "7982  11.843821 -6.402756  \n",
       "7983   9.741588 -6.433317  \n",
       "7984  10.058605 -6.755480  \n",
       "7985  10.495076 -6.194475  \n",
       "7986   9.799885 -6.237199  \n",
       "\n",
       "[7987 rows x 102 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GCNFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Wang, R. H., Ng, Y. K., Zhang, X., Wang, J., & Li, S. C. (2024). Coding genomes with gapped pattern graph convolutional network. Bioinformatics, 40(4). https://doi.org/10.1093/bioinformatics/btae188"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
